packages/slate/src/models/node.js

```diff
@@ -1982,14 +2018,34 @@ class Node {
    * @return {Node}
    */
 
-  updateNode(node) {
+  updateNode(node, willTreeChanged = true, changedChildIndex = 0) {
     if (node.key == this.key) {
       return node
     }
 
     let child = this.assertDescendant(node.key)
-    const ancestors = this.getAncestors(node.key)
 
+    let pathsCache = this.getPathsCache()
+    if (pathsCache && willTreeChanged) {
```

Comment:
We need to update paths' cache if the tree structure is changed.

We still need to traverse the whole tree if we just insert a node before `document.nodes.first()`, but we won't do that frequently.

Under most circumstances, we just need to update a sub tree's path cache.